Technical/Logging/The-Tazama-Logging-Framework.md

```diff
@@ -1 +1,405 @@
+
+An example:
+
+```js
+// No sidecar host is provided in the LoggerService constructor
+const logger = new LoggerService();
+```
+
+An alternative implementation, **opting into** logging with the sidecar (as well as [lumberjack]) may be:
+```js
+const logger = new LoggerService("192.0.0.1:1234");
+```
+> [!NOTE]  
+> If a sidecar host is provided in the `LoggerService` constructor, you have to ensure your sidecar is live on the address you provided
+
+As [gRPC] is in use, it allows Tazama to be flexible enough to allow logs coming from any client (in any programming language) as long as it adheres to the type that the server expects (defined in the protobuf format above).
+
+After receiving the logs, the [event-sidecar] sends them to [lumberjack] via [NATS]
+
+#### Lumberjack
+Lumberjack is a microservice that receives all log events before redirecting them to a central place. This microservice utilises [NATS] to listen for incoming messages with a specific subject. The [event-sidecar] sends messages with a specific subject that [lumberjack] listens for. This microservice depends on [pino-elasticsearch](https://github.com/pinojs/pino-elasticsearch), which is what acts as our transport to transform our logs from our wire format (in the protofile), to the format that [Elastic] search expects.
+
+This microservice has an additional responsibility of batching the received logs before transmission. This has a wide range of benefits as ultimately, I/O operations are reduced which will generally have implications on from improved performance coming from less network calls which can also affect costs in cloud environments.
+
+### Implementation
+The [event-sidecar] and [lumberjack] are optional components in implementation. Perhaps hardware is limited or there may be some situations whereby running more microservices is not ideal, Tazama will still have a functional logging implementation if the [event-sidecar] and [lumberjack] are excluded from deployment. As mentioned before, you can opt in to using the [event-sidecar] and [lumberjack] for logging by specifying the sidecar host when initialising the logger service.
+
+#### Configuration
+
+##### With Event-Sidecar and Lumberjack
+
+###### Lumberjack
+
+The following services are required:
+
+- [NATS] - requires a [NATS] server to be running
+- [Elastic] - log destination
+
+A sample [.env](https://github.com/frmscoe/lumberjack/blob/ce6cda49bade1a0287bc0c603330e5fbb8159455/.env.example) is provided.
+
+| Variable Name         | Purpose                               | Example             |
+|-----------------------|---------------------------------------|---------------------|
+| NATS_SERVER           | Specifies the NATS server address      | `nats://localhost:4222` |
+| NATS_SUBJECT          | Defines the NATS subject/topic name    | `Lumberjack`        |
```

Comment:
> Is the capitalization required/standard?

It does not matter, as long as the subject is valid.

> I think all our NATS subjects are now consistently lowercase

I don't think it matters. I can't check at the moment since we populate them in the build pipeline, but from a workflow: 
https://github.com/frmscoe/tms-service/blob/3bbe6e2a382cbae2dfaf00d10d9ec11d126bb895/.github/workflows/node.js.yml#L9

It seems some are still lowercase